microgen/remesh.py

```diff
@@ -0,0 +1,166 @@
+    """
+    with NamedTemporaryFile(suffix='.mesh') as boundary_triangles_file:
+        _identify_boundary_triangles_from_mesh_file(input_mesh_file, rve, boundary_triangles_file.name)
+        Mmg.mmg3d(input=boundary_triangles_file.name, output=output_mesh_file, hausd=hausd, hgrad=hgrad, hmax=hmax,
+                  hmin=hmin, hsiz=hsiz)
+
+
+def _identify_boundary_triangles_from_mesh_file(input_mesh_file: str, rve: Rve,
+                                                output_mesh_file: str) -> None:
+    """
+    Prepares the mesh file for periodic remeshing by tagging boundary triangles
+
+    This function adds a "RequiredTriangles" field
+    in the ".mesh" file that stores the boundary triangles tags.
+    mmg recognizes this field as triangles it must not remesh
+
+    :param input_mesh_file: mesh file to remesh (must be .mesh)
+    :param rve: Representative Volume Element for periodicity
+    :param output_mesh_file: output file (must be .mesh)
+    """
+    gmsh.initialize()
+    gmsh.open(input_mesh_file)
+
+    surface_triangles: list[Triangle] = _build_surface_triangles()
+    boundary_triangles_tags: list[int] = _extract_boundary_triangles_tags(surface_triangles, rve)
+
+    _write_output(boundary_triangles_tags, output_mesh_file)
+
+
+def _get_surface_triangles() -> tuple[list[int], list[int]]:
+    gmsh.model.mesh.createTopology()  # Creates boundary entities
+    (
+        _,
+        surface_triangles_tags,
+        surface_triangles_nodes_tags,
+    ) = gmsh.model.mesh.getElements(_BOUNDARY_DIM)
+    surface_triangles_nodes_tags: list[int] = list(surface_triangles_nodes_tags[0])
+    surface_triangles_tags: list[int] = list(surface_triangles_tags[0])
+
+    return surface_triangles_tags, surface_triangles_nodes_tags
+
+
+def _build_surface_triangles() -> list[Triangle]:
+    surface_triangles_tags, surface_triangles_nodes_tags = _get_surface_triangles()
+    surface_triangles_nodes_coords = _get_surface_nodes_coords(
+        surface_triangles_nodes_tags)
+
+    n_triangles = len(surface_triangles_tags)
+    surface_triangles = [
+        Triangle(
+            surface_triangles_nodes_coords[_MESH_DIM * i],
+            surface_triangles_nodes_coords[_MESH_DIM * i + 1],
+            surface_triangles_nodes_coords[_MESH_DIM * i + 2],
+            surface_triangles_tags[i],
+        )
+        for i in range(n_triangles)
+    ]
+    return surface_triangles
+
+
+def _get_surface_nodes_coords(surface_nodes: list[int]) -> np.ndarray:
+    n_nodes = len(surface_nodes)
+    surface_nodes_coords = np.zeros((n_nodes, _MESH_DIM))
+    for i in range(n_nodes):
+        surface_nodes_coords[i] = gmsh.model.mesh.getNode(surface_nodes[i])[0]
+    return surface_nodes_coords
+
+
+def _is_triangle_on_boundary(triangle: Triangle, rve: Rve) -> bool:
+    """Determines whether a triangle (defined by its 3 nodes) is on the boundary of a parallelepipedic rve"""
+
+    # there must be a better way:

```

Comment:
To avoid having to repeat the whole statement for each axis for each node, `numpy` and the `any` builtin can be helpful:

```python
def _is_triangle_on_boundary(triangle: Triangle, rve: Rve) -> bool:
    """Determines whether a triangle (defined by its 3 nodes) is on the boundary of a parallelepipedic rve"""
    coords: list[tuple[float, float, float]] = [
        (rve.x_min, rve.y_min, rve.z_min),
        (rve.x_max, rve.y_max, rve.z_max)
    ]
    nodes: list[tuple[float, float, float]] = [triangle.node1, triangle.node2, triangle.node3]
    return any(np.isclose(node, coord).all() for coord in coords for node in nodes)
```

Reply:
You must change the `np.isclose(node, coord).all()` into `np.isclose(node, coord).any()` and the tests should pass.

```python
def _is_triangle_on_boundary(triangle: Triangle, rve: Rve) -> bool:
    """Determines whether a triangle (defined by its 3 nodes) is on the boundary of a parallelepipedic rve"""
    coords = [
        (rve.x_min, rve.y_min, rve.z_min),
        (rve.x_max, rve.y_max, rve.z_max)
    ]
    nodes = triangle.node1, triangle.node2, triangle.node3
    return any(np.isclose(node, coord).any() for coord in coords for node in nodes)
```